atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/AsyncCellLoader.java

```diff
@@ -31,25 +30,29 @@
 import com.palantir.atlasdb.keyvalue.api.Cell;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.keyvalue.cassandra.async.queries.CqlQueryContext;
+import com.palantir.atlasdb.keyvalue.cassandra.async.queries.GetQuerySpec.GetQueryParameters;
+import com.palantir.atlasdb.keyvalue.cassandra.async.queries.ImmutableCqlQueryContext;
+import com.palantir.atlasdb.keyvalue.cassandra.async.queries.ImmutableGetQueryParameters;
+import com.palantir.atlasdb.keyvalue.cassandra.async.queries.ImmutableGetQuerySpec;
 import com.palantir.atlasdb.logging.LoggingArgs;
 import com.palantir.common.streams.KeyedStream;
 import com.palantir.logsafe.SafeArg;
 
-public final class AsyncCellLoader {
-
+public final class AsyncCellLoader implements AutoCloseable {
     private static final Logger log = LoggerFactory.getLogger(AsyncCellLoader.class);
 
     private final CqlClient cqlClient;
-    private final Executor executor;
+    private final ExecutorService executorService;
     private final String keyspace;
 
-    public static AsyncCellLoader create(CqlClient cqlClient, Executor executor, String keyspace) {
-        return new AsyncCellLoader(cqlClient, executor, keyspace);
+    public static AsyncCellLoader create(CqlClient cqlClient, ExecutorService executorService, String keyspace) {
+        return new AsyncCellLoader(cqlClient, executorService, keyspace);
     }
 
-    private AsyncCellLoader(CqlClient cqlClient, Executor executor, String keyspace) {
+    private AsyncCellLoader(CqlClient cqlClient, ExecutorService executorService, String keyspace) {
```

Comment:
idea: what if you don't pass ExecutorService here, but something that only exposes whatever transformation logic you need for futures. Then make CqlContext that contains CqlClient and whatever you call this new interface, and have CqlContext be AutoCloseable. Then just have one #close call.

Reply:
I now have a bit of a different interface which should in fact abstract  away how futures are combined and on which thread.